src/uu/chmod/src/chmod.rs

```diff
@@ -35,14 +35,71 @@ mod options {
     pub const FILE: &str = "FILE";
 }
 
-#[uucore::main]
-pub fn uumain(args: impl uucore::Args) -> UResult<()> {
-    let mut args = args.collect_lossy();
+// "-w -w" is the same as "-w,-w"
+// "-w file -w" is the same as "-w,-w file"
+// "-- -w file" is the same as "-w file"
+fn extract_negative_modes(mut args: impl uucore::Args) -> (Option<String>, Vec<String>) {
+    // we look up the args until "--" is found
+    // "-mode" will be extracted into parsed_cmode_vec
+    let mut parsed_cmode_vec = Vec::new();
+    let pre_double_hyphen_args: Vec<String> = args
+        .by_ref()
+        .take_while(|a| a != "--")
+        .filter_map(|arg| {
+            let arg = if let Some(arg) = arg.to_str() {
+                arg.to_string()
+            } else {
+                // FIXME: handle non-utf8 args
+                arg.to_string_lossy().to_string()
+            };
+            if arg.len() >= 2
+                && arg.starts_with('-')
+                && matches!(
+                    arg.chars().nth(1).unwrap(),
+                    'r' | 'w' | 'x' | 'X' | 's' | 't' | 'u' | 'g' | 'o' | '0'..='7'
+                )
+            {
+                parsed_cmode_vec.push(arg);
+                None
+            } else {
+                Some(arg)
+            }
+        })
+        .collect();
 
-    // Before we can parse 'args' with clap (and previously getopts),
-    // a possible MODE prefix '-' needs to be removed (e.g. "chmod -x FILE").
-    let mode_had_minus_prefix = mode::strip_minus_from_mode(&mut args);
+    let mut clean_args = Vec::new();
+    let first_after_double_hyphen = args.next();
+    let parsed_cmode = if parsed_cmode_vec.is_empty() {
+        clean_args.extend(pre_double_hyphen_args);
+        if let Some(arg) = first_after_double_hyphen {
+            // nothing found but there is an argument after "--", take it
+            clean_args.push("--".to_string());
+            clean_args.push(arg.to_string_lossy().to_string());
+            // FIXME: handle non-utf8 args
+            Some(arg.to_string_lossy().to_string())
```

Comment:
This case would also be handled by clap wouldn't it? Maybe we can remove this branch here?

Reply:
I added this path because we have to check if the previous `take_while(|a| a != "--")` has consumed the `--`.
If so, we have to re-add it to ensure clap works correctly. The (at least for me) easiest way to find out if the `--` was found+consumed is to check if there is a following argument left in the iterator. If there is a follow-up argument, then the following arg is the cmode which would be used. I also added this cmode back to the clean_args to be sure that clap doesn't fail (as this is would be the required MODE).